File: android/src/play/java/com/dooboolab/RNIap/PlayUtils.java
package com.dooboolab.RNIap;

import static com.dooboolab.RNIap.DoobooUtils.E_ALREADY_OWNED;
import static com.dooboolab.RNIap.DoobooUtils.E_DEVELOPER_ERROR;
import static com.dooboolab.RNIap.DoobooUtils.E_ITEM_UNAVAILABLE;
import static com.dooboolab.RNIap.DoobooUtils.E_NETWORK_ERROR;
import static com.dooboolab.RNIap.DoobooUtils.E_SERVICE_ERROR;
import static com.dooboolab.RNIap.DoobooUtils.E_UNKNOWN;
import static com.dooboolab.RNIap.DoobooUtils.E_USER_CANCELLED;

import android.util.Log;
import com.android.billingclient.api.BillingClient;
import com.facebook.react.bridge.Promise;

public class PlayUtils {
  private static final String TAG = "PlayUtils";
  public static final String E_PLAY_SERVICES_UNAVAILABLE = "E_PLAY_SERVICES_UNAVAILABLE";

  private static PlayUtils instance = new PlayUtils();

  public static PlayUtils getInstance() {
    return instance;
  }

  public void rejectPromiseWithBillingError(final Promise promise, final int responseCode) {
    String[] errorData = getBillingResponseData(responseCode);
    promise.reject(errorData[0], errorData[1]);
  }

  public String[] getBillingResponseData(final int responseCode) {
    String[] errorData = new String[2];
    switch (responseCode) {
      case BillingClient.BillingResponseCode.FEATURE_NOT_SUPPORTED:
        errorData[0] = E_SERVICE_ERROR;
        errorData[1] = "This feature is not available on your device.";
        break;
      case BillingClient.BillingResponseCode.SERVICE_DISCONNECTED:
        errorData[0] = E_NETWORK_ERROR;
        errorData[1] = "The service is disconnected (check your internet connection.)";
        break;
      case BillingClient.BillingResponseCode.OK:
        errorData[0] = "OK";
        errorData[1] = "";
        break;
      case BillingClient.BillingResponseCode.USER_CANCELED:
        errorData[0] = E_USER_CANCELLED;
        errorData[1] = "Payment is Cancelled.";
        break;
      case BillingClient.BillingResponseCode.SERVICE_UNAVAILABLE:
        errorData[0] = E_SERVICE_ERROR;
        errorData[1] =
            "The service is unreachable. This may be your internet connection, or the Play Store may be down.";
        break;
      case BillingClient.BillingResponseCode.BILLING_UNAVAILABLE:
        errorData[0] = E_SERVICE_ERROR;
        errorData[1] =
            "Billing is unavailable. This may be a problem with your device, or the Play Store may be down.";
        break;
      case BillingClient.BillingResponseCode.ITEM_UNAVAILABLE:
        errorData[0] = E_ITEM_UNAVAILABLE;
        errorData[1] = "That item is unavailable.";
        break;
      case BillingClient.BillingResponseCode.DEVELOPER_ERROR:
        errorData[0] = E_DEVELOPER_ERROR;
        errorData[1] = "Google is indicating that we have some issue connecting to payment.";
        break;
      case BillingClient.BillingResponseCode.ERROR:
        errorData[0] = E_UNKNOWN;
        errorData[1] = "An unknown or unexpected error has occured. Please try again later.";
        break;
      case BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED:
        errorData[0] = E_ALREADY_OWNED;
        errorData[1] = "You already own this item.";
        break;
      default:
        errorData[0] = E_UNKNOWN;
        errorData[1] = "Purchase failed with code: " + responseCode;
        break;
    }
    Log.e(TAG, "Error Code : " + responseCode);
    return errorData;
  }

  public void rejectPromisesWithBillingError(final String key, final int responseCode) {
    String[] errorData = getBillingResponseData(responseCode);
    DoobooUtils.getInstance().rejectPromisesForKey(key, errorData[0], errorData[1], null);
  }
}

File: android/src/play/java/com/dooboolab/RNIap/PlayUtils.kt
package com.dooboolab.RNIap

import android.util.Log
import com.android.billingclient.api.BillingClient
import com.facebook.react.bridge.Promise

data class BillingResponse(val code: String, val message: String)

object PlayUtils {
    fun rejectPromiseWithBillingError(promise: Promise, responseCode: Int) {
        val errorData = getBillingResponseData(responseCode)
        promise.safeReject(errorData.code, errorData.message)
    }

    fun getBillingResponseData(responseCode: Int): BillingResponse {
        val errorData= 
        when (responseCode) {
            BillingClient.BillingResponseCode.FEATURE_NOT_SUPPORTED -> {
                BillingResponse( PromiseUtils.E_SERVICE_ERROR,
                 "This feature is not available on your device.")
            }
            BillingClient.BillingResponseCode.SERVICE_DISCONNECTED -> {
                BillingResponse( PromiseUtils.E_NETWORK_ERROR
                , "The service is disconnected (check your internet connection.)")
            }
            BillingClient.BillingResponseCode.OK -> {
                BillingResponse( "OK"
                , "")
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                BillingResponse( PromiseUtils.E_USER_CANCELLED
                , "Payment is Cancelled.")
            }
            BillingClient.BillingResponseCode.SERVICE_UNAVAILABLE -> {
                BillingResponse( PromiseUtils.E_SERVICE_ERROR
                ,
                    "The service is unreachable. This may be your internet connection, or the Play Store may be down.")
            }
            BillingClient.BillingResponseCode.BILLING_UNAVAILABLE -> {
                BillingResponse( PromiseUtils.E_SERVICE_ERROR
                ,
                    "Billing is unavailable. This may be a problem with your device, or the Play Store may be down.")
            }
            BillingClient.BillingResponseCode.ITEM_UNAVAILABLE -> {
                BillingResponse( PromiseUtils.E_ITEM_UNAVAILABLE
                , "That item is unavailable.")
            }
            BillingClient.BillingResponseCode.DEVELOPER_ERROR -> {
                BillingResponse( PromiseUtils.E_DEVELOPER_ERROR
                , "Google is indicating that we have some issue connecting to payment.")
            }
            BillingClient.BillingResponseCode.ERROR -> {
                BillingResponse( PromiseUtils.E_UNKNOWN
                , "An unknown or unexpected error has occurred. Please try again later.")
            }
            BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED -> {
                BillingResponse( PromiseUtils.E_ALREADY_OWNED
                , "You already own this item.")
            }
            else -> {
                BillingResponse( PromiseUtils.E_UNKNOWN
                , "Purchase failed with code: $responseCode")
            }
        }
        Log.e(TAG, "Error Code : $responseCode")
        return errorData
    }

    fun rejectPromisesWithBillingError(key: String, responseCode: Int) {
        val error = getBillingResponseData(responseCode)
        PromiseUtils.rejectPromisesForKey(key, error.code, error.message, null)
    }
}
